torch_geometric/utils/dropout.py

```diff
@@ -208,3 +216,82 @@ def dropout_edge(edge_index: Tensor, p: float = 0.5,
         edge_mask = edge_mask.nonzero().repeat((2, 1)).squeeze()
 
     return edge_index, edge_mask
+
+
+def dropout_path(edge_index: Tensor, p: float = 0.2, walks_per_node: int = 1,
+                 walk_length: int = 3, num_nodes: Optional[int] = None,
+                 is_sorted: bool = False,
+                 training: bool = True) -> Tuple[Tensor, Tensor]:
+    r"""Drops edges from the adjacency matrix :obj:`edge_index`
+    based on random walks. The source nodes to start random walks from are
+    sampled from :obj:`edge_index` with probability :obj:`p`, following
+    a Bernoulli distribution.
+
+    The method returns (1) the retained :obj:`edge_index`, (2) the edge mask
+    indicating which edges were retained.
+
+    Args:
+        edge_index (LongTensor): The edge indices.
+        p (float, optional): Sample probability. (default: :obj:`0.2`)
+        walks_per_node (int, optional): The number of walks per node, same as
+            :class:`~torch_geometric.nn.models.Node2Vec`. (default: :obj:`1`)
+        walk_length (int, optional): The walk length, same as
+            :class:`~torch_geometric.nn.models.Node2Vec`. (default: :obj:`3`)
+        num_nodes (int, optional): The number of nodes, *i.e.*
+            :obj:`max_val + 1` of :attr:`edge_index`. (default: :obj:`None`)
+        is_sorted (bool, optional): If set to :obj:`True`, will expect
+            :obj:`edge_index` to be already sorted row-wise.
+        training (bool, optional): If set to :obj:`False`, this operation is a
+            no-op. (default: :obj:`True`)
+
+    :rtype: (:class:`LongTensor`, :class:`BoolTensor`)
+
+    Example:
+
+        >>> edge_index = torch.tensor([[0, 1, 1, 2, 2, 3],
+        ...                            [1, 0, 2, 1, 3, 2]])
+        >>> edge_index, edge_mask = dropout_path(edge_index)
+        >>> edge_index
+        tensor([[1, 2],
+                [2, 3]])
+        >>> edge_mask # masks indicating which edges are retained
+        tensor([False, False,  True, False,  True, False])
+    """
+
+    if p < 0. or p > 1.:
+        raise ValueError(f'Sample probability has to be between 0 and 1 '
+                         f'(got {p}')
+
+    num_edges = edge_index.size(1)
+    edge_mask = edge_index.new_ones(num_edges, dtype=torch.bool)
+    if not training or p == 0.0:
+        return edge_index, edge_mask
+
+    if random_walk is None:
+        raise ImportError('`dropout_path` requires `torch-cluster`.')
+
+    num_nodes = maybe_num_nodes(edge_index, num_nodes)
+    edge_orders = None
+    ori_edge_index = edge_index
+    if not is_sorted:
+        edge_orders = torch.arange(num_edges, device=edge_index.device)
+        edge_index, edge_orders = sort_edge_index(edge_index, edge_orders,
+                                                  num_nodes=num_nodes)
+
+    row, col = edge_index
+    sample_mask = torch.rand(row.size(0), device=edge_index.device) <= p
+    start = row[sample_mask].repeat(walks_per_node)
```

Comment:
Doesn't this mean we'll start many more than `walks_per_node` random walks for nodes with a high out-degree.
Sorry for the back and forth, but I felt the way you defined start perviously was better and actually meant we start exactly `walks_per_node` random walks`.

Reply:
Agreed. Or maybe we can add an option such as `walk_by='edge'` or `walk_by='node'` to align them?

Reply:
Maybe I am missing something. Why do we want to support `walk_by='edge'`?

Reply:
Oh sorry I meant we can sample starting nodes node-wise (previous version) or edge-wise (current version).  For edge-wise sampling, as you mentioned, we can start random walks for nodes with a high out-degree, this seems to make sense for graphs with particular imbalanced structures.

Reply:
I think this is the last one that needs to be resolved.

Reply:
No strong opinion. I think we can safely merge this and wait on user feedback. To me, it makes sense to sample per edge since we are dropping paths.

Reply:
Maybe just add a comment for now?

Reply:
> Maybe just add a comment for now and clarify in the doc?
Good idea.

Reply:
Alternatively, we can let `p` accept `Tensor` to specify the nodes to start random walks, rather than randomly sampled from nodes and edges. Let's leave it as future work and merge it now.

Reply:
Apologies, i forgot to respond. its good that you merged it :).

Reply:
No worries :)